src/protocols/secure_channel/SimpleSessionResumptionStorage.cpp

```diff
@@ -0,0 +1,273 @@
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
+    FabricIndex fabricIndex;
+    ReturnErrorOnFailure(reader.Next(kFabricIndexTag));
+    ReturnErrorOnFailure(reader.Get(fabricIndex));
+
+    NodeId peerNodeId;
+    ReturnErrorOnFailure(reader.Next(kPeerNodeIdTag));
+    ReturnErrorOnFailure(reader.Get(peerNodeId));
+
+    ReturnErrorOnFailure(reader.ExitContainer(containerType));
+    ReturnErrorOnFailure(reader.VerifyEndOfContainer());
+
+    node = ScopedNodeId(peerNodeId, fabricIndex);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSessionResumptionStorage::DeleteLink(ConstResumptionIdView resumptionId)
+{
+    DefaultStorageKeyAllocator keyAlloc;
+    ReturnErrorOnFailure(mStorage->SyncDeleteKeyValue(StorageKey(keyAlloc, resumptionId)));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSessionResumptionStorage::SaveState(const ScopedNodeId & node, ConstResumptionIdView resumptionId,
+                                                     const Crypto::P256ECDHDerivedSecret & sharedSecret, const CATValues & peerCATs)
+{
+    // Save session state into key: /f/<fabricIndex>/s/<nodeId>
+    uint8_t buf[MaxStateSize()];
+
+    TLV::TLVWriter writer;
+    writer.Init(buf);
+
+    TLV::TLVType outerType;
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, outerType));
+
+    ReturnErrorOnFailure(writer.Put(kResumptionIdTag, resumptionId));
+
+    ReturnErrorOnFailure(writer.Put(kSharedSecretTag, ByteSpan(sharedSecret.ConstBytes(), sharedSecret.Length())));
+
+    CATValues::Serialized cat;
+    peerCATs.Serialize(cat);
+    ReturnErrorOnFailure(writer.Put(kCATTag, ByteSpan(cat)));
+
+    ReturnErrorOnFailure(writer.EndContainer(outerType));
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrDie(CanCastTo<uint16_t>(len));
+
+    DefaultStorageKeyAllocator keyAlloc;
+    ReturnErrorOnFailure(mStorage->SyncSetKeyValue(StorageKey(keyAlloc, node), buf, static_cast<uint16_t>(len)));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSessionResumptionStorage::LoadState(const ScopedNodeId & node, ResumptionIdStorage & resumptionId,
+                                                     Crypto::P256ECDHDerivedSecret & sharedSecret, CATValues & peerCATs)
+{
+    uint8_t buf[MaxStateSize()];
+    uint16_t len = static_cast<uint16_t>(MaxStateSize());
+
+    DefaultStorageKeyAllocator keyAlloc;
+    ReturnErrorOnFailure(mStorage->SyncGetKeyValue(StorageKey(keyAlloc, node), buf, len));
+
+    TLV::ContiguousBufferTLVReader reader;
+    reader.Init(buf, len);
+
+    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag()));
+    TLV::TLVType containerType;
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
+    ByteSpan resumptionIdSpan;
+    ReturnErrorOnFailure(reader.Next(kResumptionIdTag));
+    ReturnErrorOnFailure(reader.Get(resumptionIdSpan));
+    std::copy(resumptionIdSpan.begin(), resumptionIdSpan.end(), resumptionId.begin());
```

Comment:
This is a buffer overrun if it happens that `resumptionIdSpan.size() > resumptionIdSpan`, and nothing is ensuring it's not.